cc/client/client_test.cc

```diff
@@ -74,7 +75,7 @@ class TestTransport : public TransportWrapper {
   }
 
  private:
-  KeyPair key_pair_;
+  std::shared_ptr<EncryptionKeyProvider> encryption_key_provider_;
```

Comment:
Generally std::shared_ptr are not encouraged within Google (other than a few edge cases). Can we use std::unique_ptr?

Reply:
Encryption key provider will be shared with every instance of `Encryptor`. I'm not sure whether we should copy key provider every time, because it'll end up with copying the private key.

Reply:
If you have to use a shared pointer, the shared object must not be mutable (otherwise you need to protect it with a mutex), so at the very least this should be `std::shared_ptr<const EncryptionKeyProvider> encryption_key_provider_;`.

@andrisaar any thoughts on whether this is appropriate use of a shared pointer, or whether there are any other preferred options that avoid copying the private key?

Reply:
Not sure it could be constant. `EncryptionKeyProvider` implements the `RecipientContextGenerator` interface, and this interface will be also implemented by an Orchestrator client (which is not `const`). And what currently ends up inside an `Encryptor` is a `shared_ptr<RecipientContextGenerator>` and not the `shared_ptr<EncryptionKeyProvider>`.

Reply:
You would then have to protect it with a mutex to avoid multiple concurrent mutations.

Reply:
I guess the mutex will have to be in the RecipientContextGenerator to make it thread safe. If the intention is to store a RecipientContextGenerator, then that should probably already be indicated by the field type.

As a separate point on mutability: I know in Rust Tonic clients are mutable, but does the same also apply to C++ clients? I thought it was only the context that had to be mutable, and you need a new context for every request anyway.

Reply:
Eyeballing this code I don't see why it needs to be a `shared_ptr`.

The EncryptionKeyProvider is constructed in the `TestTransport` constructor. There are no getters or setters, so the only way it is ever used is to pass it to ServerEncryptor in `Invoke`. The lifetime of said ServerEncryptor is wholly contained within the body of Invoke, so the EncryptionKeyProvider trivially outlives the ServerEncryptor.
Thus, the ServerEncryptor should take a `EncryptorKeyProvider&` instead of a shared_ptr.

Are there any situations where a ServerEncryptor could outlive the provider?

Reply:
> Are there any situations where a ServerEncryptor could outlive the provider?

`ServerEncryptor` will never outlive the provider, because it needs it to generate encryption contexts

Reply:
Updated to take in a reference, since `RecipientContextGenerator` argument is a long-term object containing the private key and it should outlive the per-session `ServerEncryptor` object.
Also added a comment.